src/compiler/moduleNameResolver.ts

```diff
@@ -637,10 +652,59 @@ namespace ts {
 
         function getOrCreateCacheForDirectory(directoryName: string, redirectedReference?: ResolvedProjectReference) {
             const path = toPath(directoryName, currentDirectory, getCanonicalFileName);
-            return getOrCreateCache<ESMap<string, T>>(directoryToModuleNameMap, redirectedReference, path, () => new Map());
+            return getOrCreateCache<ModeAwareCache<T>>(directoryToModuleNameMap, redirectedReference, path, () => createModeAwareCache());
         }
     }
 
+    /* @internal */
+    export function createModeAwareCache<T>(): ModeAwareCache<T> {
+        const underlying = new Map<string, T>();
+        const memoizedReverseKeys = new Map<string, [specifier: string, mode: ModuleKind.CommonJS | ModuleKind.ESNext | undefined]>();
+
+        const cache: ModeAwareCache<T> = {
+            get(specifier, mode) {
+                return underlying.get(getUnderlyingCacheKey(specifier, mode));
+            },
+            set(specifier, mode, value) {
+                underlying.set(getUnderlyingCacheKey(specifier, mode), value);
+                return cache;
+            },
+            delete(specifier, mode) {
+                underlying.delete(getUnderlyingCacheKey(specifier, mode));
+                return cache;
+            },
+            has(specifier, mode) {
+                return underlying.has(getUnderlyingCacheKey(specifier, mode));
+            },
+            forEach(cb) {
+                return underlying.forEach((elem, key) => {
+                    const [specifier, mode] = memoizedReverseKeys.get(key)!;
+                    return cb(elem, specifier, mode);
+                });
+            },
+            size() {
+                return underlying.size;
+            }
+        };
+        return cache;
+
+        function getUnderlyingCacheKey(specifier: string, mode: ModuleKind.CommonJS | ModuleKind.ESNext | undefined) {
+            const result = mode === undefined ? specifier : `${mode}|${specifier}`;
+            memoizedReverseKeys.set(result, [specifier, mode]);

```

Comment:
No? Then I'd need the compiler options in the cache key/arguments and I'd need to take that in as a parameter and switch on that and... It doesn't really change much. Plus, I really don't think this is going to matter much - this doesn't increase the complexity of the operation at all (there's no extra iteration being done) and only increases the memory usage by the size of a few pointers per element. This is highly unlikely to be any kind of bottleneck, imo.

Reply:
> Then I'd need the compiler options in the cache key

The cache can be used across different programs?

Reply:
I should be able to test this with `--extendedDiagnostics`... I don’t think our perf suite reports module resolution time (or program construction time) directly, IIRC.

Reply:
> The cache can be used across different programs?

Right, exactly - I'd be trading the one thing I needed a reverse lookup cache for for two (since it's not like I can guarantee the new resolution modes won't ever be used in a context prior to cache construction). It wouldn't really help - the key is already minimal.